File: src/main/java/org/wysko/midis2jam2/instrument/percussive/Woodblocks.java
package org.wysko.midis2jam2.instrument.percussive;

import com.jme3.math.Quaternion;
import com.jme3.math.Vector3f;
import com.jme3.scene.Node;
import com.jme3.scene.Spatial;
import org.jetbrains.annotations.NotNull;
import org.wysko.midis2jam2.Midis2jam2;
import org.wysko.midis2jam2.instrument.TwelveDrumOctave;
import org.wysko.midis2jam2.instrument.percussion.drumset.PercussionInstrument;
import org.wysko.midis2jam2.midi.MidiChannelSpecificEvent;

import java.util.List;
import java.util.stream.IntStream;

import static com.jme3.math.FastMath.HALF_PI;
import static org.wysko.midis2jam2.Midis2jam2.rad;

public class Woodblocks extends TwelveDrumOctave {
	
	Node[] woodBlockNodes = new Node[12];
	
	/**
	 * @param context   the context to the main class
	 * @param eventList the event list
	 */
	public Woodblocks(@NotNull Midis2jam2 context,
	                  @NotNull List<MidiChannelSpecificEvent> eventList) {
		super(context, eventList);
		
		IntStream.range(0, 12).forEach(i -> woodBlockNodes[i] = new Node());
		
		
		for (int i = 0; i < 12; i++) {
			malletNodes[i] = new Node();
			Spatial child = context.loadModel("DrumSet_Stick.obj", "StickSkin.bmp");
			child.setLocalTranslation(0, 0, -5);
			malletNodes[i].setLocalTranslation(0, 0, 18);
			malletNodes[i].attachChild(child);
			Node oneBlock = new Node();
			oneBlock.attachChild(malletNodes[i]);
			Woodblock woodblock = new Woodblock(i);
			decayeds[i] = woodblock;
			oneBlock.attachChild(woodblock.highestLevel);
			woodBlockNodes[i].attachChild(oneBlock);
			oneBlock.setLocalTranslation(0, 0, 20);
			woodBlockNodes[i].setLocalRotation(new Quaternion().fromAngles(0, rad(7.5 * i), 0));
			woodBlockNodes[i].setLocalTranslation(0, 0.3f * i, 0);
			instrumentNode.attachChild(woodBlockNodes[i]);
		}
		
		instrumentNode.setLocalTranslation(75, 0, -35);
		
	}
	
	@Override
	public void tick(double time, float delta) {
		super.tick(time, delta);
		for (TwelfthOfOctaveDecayed woodblock : decayeds) {
			woodblock.tick(time, delta);
		}
	}
	
	@Override
	protected void moveForMultiChannel() {
		offsetNode.setLocalTranslation(0, 15 + 3.6f * indexForMoving(), 0);
		instrumentNode.setLocalRotation(new Quaternion().fromAngles(0, -HALF_PI + HALF_PI * indexForMoving(), 0));
	}
	
	public class Woodblock extends TwelveDrumOctave.TwelfthOfOctaveDecayed {
		
		public Woodblock(int i) {
			Spatial mesh = context.loadModel("WoodBlockSingle.obj", "SimpleWood.bmp");
			mesh.setLocalScale(1 - 0.036363636f * i);
			animNode.attachChild(mesh);
			
		}
		
		public void tick(double time, float delta) {
			Vector3f localTranslation = highestLevel.getLocalTranslation();
			if (localTranslation.y < -0.0001) {
				highestLevel.setLocalTranslation(0, Math.min(0, localTranslation.y + (PercussionInstrument.DRUM_RECOIL_COMEBACK * delta)), 0);
			} else {
				highestLevel.setLocalTranslation(0, 0, 0);
			}
		}
	}
}

File: src/main/java/org/wysko/midis2jam2/instrument/family/percussive/Woodblocks.kt
/*
 * Copyright (C) 2021 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */
package org.wysko.midis2jam2.instrument.family.percussive

import com.jme3.math.FastMath
import com.jme3.math.Quaternion
import com.jme3.scene.Node
import org.wysko.midis2jam2.Midis2jam2
import org.wysko.midis2jam2.instrument.family.percussion.drumset.PercussionInstrument
import org.wysko.midis2jam2.midi.MidiChannelSpecificEvent
import org.wysko.midis2jam2.util.Utils.rad

/** The Woodblocks consist of 12 blocks. */
class Woodblocks(context: Midis2jam2, eventList: List<MidiChannelSpecificEvent>) :
    OctavePercussion(context, eventList) {

    override fun tick(time: Double, delta: Float) {
        super.tick(time, delta)
        twelfths.forEach { it!!.tick(delta) }
    }

    override fun moveForMultiChannel(delta: Float) {
        val index = updateInstrumentIndex(delta)
        offsetNode.setLocalTranslation(0f, 15 + 3.6f * index, 0f)
        instrumentNode.localRotation =
            Quaternion().fromAngles(0f, -FastMath.HALF_PI + FastMath.HALF_PI * index, 0f)
    }

    /** A single Woodblock. */
    inner class Woodblock(i: Int) : TwelfthOfOctaveDecayed() {
        override fun tick(delta: Float) {
            val localTranslation = highestLevel.localTranslation
            if (localTranslation.y < -0.0001) {
                highestLevel.setLocalTranslation(
                    0f,
                    0f.coerceAtMost(localTranslation.y + PercussionInstrument.DRUM_RECOIL_COMEBACK * delta), 0f
                )
            } else {
                highestLevel.setLocalTranslation(0f, 0f, 0f)
            }
        }

        init {
            val mesh = context.loadModel("WoodBlockSingle.obj", "SimpleWood.bmp")
            mesh.setLocalScale(1 - 0.036f * i)
            animNode.attachChild(mesh)
        }
    }

    init {
        for (i in 0..11) {
            /* Load drum stick */
            val stick = context.loadModel("DrumSet_Stick.obj", "StickSkin.bmp")
            stick.setLocalTranslation(0f, 0f, -5f)

            /* Initialize node and attach stick */malletNodes[i] = Node()
            malletNodes[i].setLocalTranslation(0f, 0f, 18f)
            malletNodes[i].attachChild(stick)
            val oneBlock = Node()
            oneBlock.attachChild(malletNodes[i])
            val woodblock = Woodblock(i)
            twelfths[i] = woodblock
            oneBlock.attachChild(woodblock.highestLevel)
            percussionNodes[i].attachChild(oneBlock)
            oneBlock.setLocalTranslation(0f, 0f, 20f)
            percussionNodes[i].localRotation = Quaternion().fromAngles(0f, rad(7.5 * i), 0f)
            percussionNodes[i].setLocalTranslation(0f, 0.3f * i, 0f)
            instrumentNode.attachChild(percussionNodes[i])
        }
        instrumentNode.setLocalTranslation(75f, 0f, -35f)
    }
}